### DISCUSSION TITLE
antsRegistrationSyNQuick.sh running out of memory

### DISCUSSION THREAD

**Discussion by mrinmayik** [ANTsX/ANTs#1582] [Q&A] [answered]
Hello,

I'm trying to register time-of-flight (TOF) angiography scans with T1s collected at 3T and 7T on a high-performance computing cluster.

The resolution of the TOF scans is 0.28 mm isotropic, and that of the T1s is 0.7mm isotropic. I first tried to register the TOF to the raw T1, but the TOF was getting downsampled to the T1 resolution. To get around this, I resampled the T1 to the TOF resolution using `ResampleImage`, and am now registering the TOF to the resampled T1 (this was suggested in an ANTs message board post).

However, antsRegistrationSyNQuick.sh is running out of memory. I have tried allocating 20g of memory. I have set `ITK_GLOBAL_DEFAULT_NUMBER_OF_THREADS` to 24, and in the output I can see that the process is using 24 threads. 
```
Dimensionality:           3
 Output name prefix:       /home/mrinmayi/projects/def-rolsen/mrinmayi/Data/HippocampalArtery/derivatives//AlignedTOF_ses-7T_2021/sub-001/sub-001_ses-7T_2021_part-compressed_coregTOFtoT1SynQuick_
 Fixed images:             /home/mrinmayi/projects/def-rolsen/mrinmayi/Data/HippocampalArtery/derivatives//AlignedTOF_ses-7T_2021/sub-001/sub-001_part-compressed_T1Resampled.nii
 Moving images:            /home/mrinmayi/projects/def-rolsen/mrinmayi/Data/HippocampalArtery/rawdata//sub-001/ses-7T_2021//angio/sub-001_ses-7T_2021_run-01_part-compressed_angio.nii
 Mask images:              
 Initial transforms:       
 Number of threads:        24
 Spline distance:          26
 Transform type:           s
 MI histogram bins:        32
 Precision:                d
 Use histogram matching    0
```

This is the command I'm using (all defaults):
```
antsRegistrationSyNQuick.sh -d 3 \
            -f ${f} \
            -m ${m} \
            -o ${TOFOutputDir}/${sid}_${session}_${scanpart}_coregTOFtoT1SynQuick_
```

Do you have any suggestions for how I might get this to work?

Thank you,
Mrinmayi

**Comment by cookpa** [comment, marked as answer]
Do you need the registration to happen at the 0.28mm resolution? Or you just want your deformed image to have that spacing? 

If the latter, you can do the registration using the T1w as the fixed image, then separately run `antsApplyTransforms` with your upsampled T1w as the reference image (`-r`), and you will get deformed TOF images at the higher resolution.

If you're aligning different modalities collected on the same brain, a lower-resolution warp, global affine, or even a rigid transform might be sufficient. Are there non-rigid distortions between the images? How detailed are they? It might be possible to well approximate them with a lower resolution deformation field, which will save memory.

General ways to save memory:

* Use `-p f` with the `antsRegistrationSyN*.sh` scripts to reduce memory by using single-precision floating point numbers in the registration.
* Reduce the maximum resolution of the warps (using `-f` with `antsRegistration`).
* Reduce the number of threads. I would set this to no more than 8, as the performance gains tend to decrease anyway beyond this point.
* Crop the images, if possible. Sometimes the image contains a lot of extra space, either background or uninteresting anatomy. Reducing the size of the input images will reduce memory use, but some care is needed to keep the FOV consistent between the images. I usually crop then pad with 10 voxels of empty space (`ImageMath` will do this with `PadImage`).

**Reply by mrinmayik** [reply]
Thanks a lot for your response!

> If the latter, you can do the registration using the T1w as the fixed image, then separately run antsApplyTransforms with your upsampled T1w as the reference image (-r), and you will get deformed TOF images at the higher resolution.

Do you mean that I can use `antsRegistrationSyNQuick.sh` to calculate the transform from the TOF to the low-res T1, and the apply that transform in `antsApplyTransforms` with upsampled T1 as the reference image? Is the following command correct (i.e., should I be using the transformation in the `*_0GenericAffine.mat` file?
```
antsApplyTransforms --transform OutputFile_0GenericAffine.mat \
                -n NearestNeighbor \
                -o HiResTOF_Warped.nii \
                -d 3 \
                -i HiResTOF_Raw.nii \
                -r UpsampledT1.nii
```

> If you're aligning different modalities collected on the same brain, a lower-resolution warp, global affine, or even a rigid transform might be sufficient. Are there non-rigid distortions between the images? How detailed are they? It might be possible to well approximate them with a lower resolution deformation field, which will save memory.

Yes, they're images of the same brain, but across modalities and sometimes across sessions and field strengths (e.g., aligning a 3T TOF with a 7T T1). Should I use `antsRegistrationSyNQuick.sh` with the transform (`-t`) set to `a` (rigid + affine (2 stages))?

Thanks a lot for your suggestions!
Mrinmayi

**Reply by cookpa** [reply]
That transform command looks correct to apply the affine component only. I would add `--verbose` and check there's no errors and that everything looks sensible.

You can make the reference image any image in the same *physical space* as the fixed image in the registration. That means you are free to use `ResampleImage` on your fixed image and make it have larger or smaller voxels.

For more on applying warps, please see

https://github.com/ANTsX/ANTs/wiki/Forward-and-inverse-warps-for-warping-images,-pointsets-and-Jacobians

Regarding the transform to use, you can use `antsApplyTransforms` to apply the affine component only (as in your example) and then again with both the affine + warp.nii.gz (see the wiki page for the correct way to do this, order matters). This will let you evaluate how much the solution changes between the affine and deformable stages.

**Reply by mrinmayik** [reply]
Got it - that works! Thanks so much!
Mrinmayi
